pkg/gpt/gpt.go

```diff
@@ -0,0 +1,224 @@
+	}
+	if h := crc32.ChecksumIEEE(partBlocks); h != g.PartCRC {
+		return g, fmt.Errorf("%s Partition CRC: Header %v, computed checksum is %08x, header has %08x", which, g, h, g.PartCRC)
+	}
+
+	hdr := make([]byte, g.HeaderSize)
+	n, err = r.ReadAt(hdr, off)
+	if n != len(hdr) || err != nil {
+		return nil, fmt.Errorf("%s Reading Header: Wanted %d bytes, got %d: %v", which, n, len(hdr), err)
+	}
+	// Zap the checksum in the header to 0.
+	copy(hdr[16:], []byte{0, 0, 0, 0})
+	if h := crc32.ChecksumIEEE(hdr); h != g.CRC {
+		return g, fmt.Errorf("%s Header CRC: computed checksum is %08x, header has %08x", which, h, g.CRC)
+	}
+
+	// Now read in the partition table entries.
+	g.Parts = make([]Part, g.NPart)
+
+	for i := int64(0); i < int64(g.NPart); i++ {
+		if err := binary.Read(io.NewSectionReader(r, int64(g.PartStart*BlockSize)+i*s, s), binary.LittleEndian, &g.Parts[i]); err != nil {
+			return nil, fmt.Errorf("%s GPT partition %d failed: %v", which, i, err)
+		}
+	}
+
+	return g, nil
+
+}
+
+// Write writes the GPT to w, both primary and backup. It generates the CRCs before writing.
+// It takes an io.Writer and assumes that you are correctly positioned in the output stream.
+// This means we must adjust partition numbers by subtracting one from them.
+func Write(w io.WriterAt, g *GPT) error {
+	ps := int64(g.PartStart * BlockSize)
+	// The maximum extent is ps + NPart * PartSize
+	var h = make([]byte, uint64(g.NPart*g.PartSize))
+	s := int64(g.PartSize)
+	for i := int64(0); i < int64(g.NPart); i++ {
+		var b bytes.Buffer
+		if err := binary.Write(&b, binary.LittleEndian, &g.Parts[i]); err != nil {
+			return err
+		}
+		copy(h[i*s:], b.Bytes())
+	}
+	if _, err := w.WriteAt(h, ps); err != nil {
+		return fmt.Errorf("Writing %d bytes of partition table at %v: %v", len(h), ps, err)
+	}
+
+	g.PartCRC = crc32.ChecksumIEEE(h[:])
+	g.CRC = 0
+	var b bytes.Buffer
+	if err := binary.Write(&b, binary.LittleEndian, &g.Header); err != nil {
+		return err
+	}
+	h = make([]byte, g.HeaderSize)
+	copy(h, b.Bytes())
+	g.CRC = crc32.ChecksumIEEE(h[:])
+	b.Reset()
+	if err := binary.Write(&b, binary.LittleEndian, g.CRC); err != nil {
+		return err
+	}
+	copy(h[16:], b.Bytes())
+
+	_, err := w.WriteAt(h, int64(g.CurrentLBA * BlockSize))
+	return err
+}
+
+// New reads in the primary and backup GPT from a disk and returns a pointer to them.
+// There are some checks it can apply. It can return with a
+// one or more headers AND an error. Sorry. Experience with some real USB sticks
+// is showing that we need to return data even if there are some things wrong.
+func New(r io.ReaderAt) (*GPT, *GPT, error) {
+	g, err := Table(r, HeaderOff)
+	// If we can't read the table it's kinda hard to find the backup.
+	// Bit of a flaw in the design, eh?
+	if err != nil {
+		return nil, nil, err
+	}
+
+	b, err := Table(r, int64(g.BackupLBA*BlockSize))
+	if err != nil {
+		// you go to war with the army you have
```

Comment:
maybe you should print out the error here.

Reply:
?? I do. What am I missing?

Reply:
oh I see, sorry, I used to print it out but it actually gets pretty annoying to have packages do prints like that. I removed it.